ext/bcmath/libbcmath/src/div.c

```diff
@@ -31,219 +31,403 @@
 
 #include "bcmath.h"
 #include "private.h"
+#include "convert.h"
 #include <stdbool.h>
 #include <stddef.h>
 #include <string.h>
 #include "zend_alloc.h"
 
+#define BC_VECTOR_STANDARD_DIV_CARRY_LIMIT BC_VECTOR_NO_OVERFLOW_ADD_COUNT
 
-/* Some utility routines for the divide:  First a one digit multiply.
-   NUM (with SIZE digits) is multiplied by DIGIT and the result is
-   placed into RESULT.  It is written so that NUM and RESULT can be
-   the same pointers.  */
+static const BC_VECTOR POW_10_LUT[8] = {
+	1, 10, 100, 1000, 10000, 100000, 1000000, 10000000
+};
 
-static void _one_mult(unsigned char *num, size_t size, int digit, unsigned char *result)
+/*
+ * Use this function when the number of arrays in n2 is 1.
+ * In this case, there is no need to add back.
+ */
+static void bc_fast_div(BC_VECTOR *n1_vector, size_t n1_arr_size, BC_VECTOR n2_vector, BC_VECTOR *quot_vector, size_t quot_arr_size)
 {
-	size_t carry, value;
-	unsigned char *nptr, *rptr;
+	size_t n1_top_index = n1_arr_size - 1;
+	size_t quot_top_index = quot_arr_size - 1;
+	for (size_t i = 0; i < quot_arr_size - 1; i++) {
+		if (n1_vector[n1_top_index - i] < n2_vector) {
+			quot_vector[quot_top_index - i] = 0;
+			n1_vector[n1_top_index - i - 1] += n1_vector[n1_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+			continue;
+		}
+		quot_vector[quot_top_index - i] = n1_vector[n1_top_index - i] / n2_vector;
+		n1_vector[n1_top_index - i] -= n2_vector * quot_vector[quot_top_index - i];
+		n1_vector[n1_top_index - i - 1] += n1_vector[n1_top_index - i] * BC_VECTOR_BOUNDARY_NUM;
+		n1_vector[n1_top_index - i] = 0;
+	}
+	/* last */
+	quot_vector[0] = n1_vector[0] / n2_vector;
+}
 
-	if (digit == 0) {
-		memset(result, 0, size);
+/*
+ * This is a function that uses the 3N-digit/2N-digit subalgorithm of the Burnikel-Ziegler division algorithm.
+ * When calculating a temporary quotient, it may be necessary to handle large values that cannot be handled by
+ * BC_VECTOR, in which case it is used as divide-and-conquer.
```

Comment:
I learned this algorithm from pages 5 and 6 of the following PDF:

https://pure.mpg.de/rest/items/item_1819444_4/component/file_2599480/content

Please note that there is a typographical error, `A4` is written where `A3` should be written.

I couldn't find any other good links but I'm not sure if I should leave this as a comment.